libs/go/aealite/protocols/dialogue.go

```diff
@@ -0,0 +1,437 @@
+}
+
+func (dialogue Dialogue) basicValidationNonInitialMessage(message AbstractMessage) (bool, string) {
+	dialogue_reference := message.dialogueReference
+	if dialogue_reference[0] != dialogue.dialogueLabel.dialogueReference[0] {
+		return false, "Invalid dialogue_reference[0]."
+	}
+	err := dialogue.validateMessageId(message)
+	if err != nil {
+		return false, err
+	}
+	err = dialogue.validateMessageTarget(message)
+	if err != nil {
+		return false, err
+	}
+	return true, "The non-initial message passes basic validation."
+}
+
+func (dialogue Dialogue) validateMessageTarget(message AbstractMessage) string {
+	target = message.target
+    performative = message.performative
+
+    if message.messageId == StartingMessageId {
+		if target == StartingTarget {
+			return nil
+		}
+        return "Invalid target."
+	}
+    if message.messageId != StartingMessageId && target == StartingTarget {
+		return "Invalid target. Expected a non-zero integer."
+	}
+    var latest_ids []int
+    if ok, id := dialogue.lastIncomingMessage(); ok {
+		latest_ids = append(latest_ids, id)
+	}
+	if ok, id := dialogue.lastOutgoingMessage(); ok {
+		latest_ids = append(latest_ids, id)
+	}
+    if target > max(latest_ids) {
+        return "Invalid target."
+	}
+
+	// TODO
+	// implement function getmessageby id
+	target_message := dialogue.getMessageById(target)
+    
+	if target_message == nil {
+		return "Invalid target"
+	}
+    target_performative = target_message.performative
+    if performative not in self.rules.get_valid_replies(target_performative):
+        return "Invalid performative. Expected one of {}. Found {}.".format(
+            self.rules.get_valid_replies(target_performative), performative
+        )
+    return nil
+}
+
+func max(list []int) int {
+	max := list[0]
+    for i :=1; i < len(s); i++ {
+        if max < list[i] {
+            max = list[i]
+        }
+    }
+	return max
+}
+
+func (dialogue Dialogue) validateMessageId(message AbstractMessage) string {
+	var next_message_id int
+	if message.to != dialogue.selfAddress {
+		next_message_id = dialogue.getOutgoingNextMessageId()
+	} else {
+		next_message_id = dialogue.getIncomingNextMessageId()
+	}
+    if message.message_id != next_message_id {
+		return "Invalid message_id."
+	}
+    return nil
+}
+
+func (dialogue Dialogue) getOutgoingNextMessageId() int {
+	next_message_id := StartingMessageId
+    if dialogue.lastOutgoingMessage() {
+		next_message_id = dialogue.lastMessageId + 1
+	}
+    if dialogue.isSelfInitiated() {
+		next_message_id = 0 - next_message_id
+	}
+	return next_message_id
+}
+
+func (dialogue Dialogue) getIncomingNextMessageId() {
+	next_message_id := StartingMessageId
+    if dialogue.lastIncomingMessage(){
+		next_message_id = dialogue.lastMessageId + 1
+	}
+    if dialogue.isSelfInitiated() {
+		next_message_id = 0 - next_message_id
+	}
+	return next_message_id
+}
+
+func (dialogue Dialogue)lastOutgoingMessage() (bool, int){
```

Comment:
I would use `*AbstractMessage`, the closest type to Python's `Optional[Message]`